### PR TITLE
feat: add permit2 entrypoints to the periphery

### PR DESCRIPTION
**Pull request by bmzig**
This adds two more functions to the spoke pool periphery contract: `depositWithPermit2` and `swapAndBridgeWithPermit2`. Both use the signature transfer logic in permit2. 

There is an assumption that the amount to spend on the permit2 signature is equal to the amount to swap (for a swap and bridge tx) or the amount to deposit (for a deposit tx). I believe that this assumption is correct, since a user should not give the periphery an allowance, so it should only ever `transferFrom` the amount specified in the permit details anyways.

### CODE REVIEW

**Review by nicholaspai — COMMENTED** [review]
This looks good. Can you add a forge unit test for swapAndBridgeWithPermit2?

File: contracts/SpokePoolV3Periphery.sol
Comment (nicholaspai) [inline contracts/SpokePoolV3Periphery.sol:178]: Maybe we should add a check that permit2 is initialized to a real contract similar to the spoke pool address check?

File: contracts/external/interfaces/IPermit2.sol
Comment (nicholaspai) [inline contracts/external/interfaces/IPermit2.sol:21]: Do we not want to use the permitWitnessTransferFrom function? I'm not sure myself, just asking the question. Maybe @dohaki would know better

File: test/evm/foundry/local/SpokePoolPeriphery.t.sol
Comment (nicholaspai) [inline test/evm/foundry/local/SpokePoolPeriphery.t.sol:59]:
```suggestion
        permit2 = IPermit2(vm.addr(4));
```

File: contracts/SpokePoolV3Periphery.sol
Comment (bmzig) [inline contracts/SpokePoolV3Periphery.sol:653]: Can we actually just always return true? My reasoning is that this contract can already perform arbitrary calls -- it should _never_ have tokens or approvals -- and presumably the only thing this is used for is to transfer funds from this contract to an exchange while in the middle of a swap transaction.

File: contracts/external/interfaces/IPermit2.sol
Comment (bmzig) [reply contracts/external/interfaces/IPermit2.sol]: You're right. We do want it https://github.com/across-protocol/contracts/pull/782/commits/4f3b2d6b094548efeeb11dacab40acd11f67bccd

File: contracts/SpokePoolV3Periphery.sol
Comment (bmzig) [inline contracts/SpokePoolV3Periphery.sol:707]: Seems very strange, but we should also just be able to make these all just dummy values, since we never read anything from permit2, which is the contract that uses these values to process the approval. Sounds odd though.

File: contracts/SpokePoolV3Periphery.sol
Comment (nicholaspai) [inline contracts/SpokePoolV3Periphery.sol:188]: Shouldn't we be enforcing that the transferDetails.to == address(this)? We only want to pull funds into this contract, right?

File: contracts/SpokePoolV3Periphery.sol
Comment (nicholaspai) [inline contracts/SpokePoolV3Periphery.sol:313]: same question as https://github.com/across-protocol/contracts/pull/782/files#r1868191708 about `transferDetails`

File: contracts/SpokePoolV3Periphery.sol
Comment (nicholaspai) [inline contracts/SpokePoolV3Periphery.sol:715]: so this sigDeadline enforces that this permit is atomically used right?

File: contracts/SpokePoolV3Periphery.sol
Comment (nicholaspai) [reply contracts/SpokePoolV3Periphery.sol:653]: i think you're correct in checking `expectingPermit2Callback`, but want to see what others think

**Review by nicholaspai — COMMENTED** [review]
This looks right to me, just have a couple q's

File: contracts/SpokePoolV3Periphery.sol
Comment (bmzig) [reply contracts/SpokePoolV3Periphery.sol:715]: Correct: https://github.com/Uniswap/permit2/blob/cc56ad0f3439c502c246fc5cfcc3db92bb8b7219/src/AllowanceTransfer.sol#L34

File: contracts/SpokePoolV3Periphery.sol
Comment (bmzig) [reply contracts/SpokePoolV3Periphery.sol:188]: We can, but if somebody did this, I think it means that they just passed in some random permit signature. The function would revert eventually, since the periphery would have no funds to swap or deposit.

File: contracts/SpokePoolV3Periphery.sol
Comment (bmzig) [inline contracts/SpokePoolV3Periphery.sol:172]: This will be updated to a new ACROSS_SWAP_AND_DEPOSIT_DATA_TYPE in a follow-up PR (along with the other swapAndBridge functions) so that the user can't get destroyed if a malicious person modified the swapData to accept a low output amount (and thus a low inputAmount for the across input token).

File: contracts/SpokePoolV3Periphery.sol
Comment (mrice32) [inline contracts/SpokePoolV3Periphery.sol:171]: Permit2 cannot be approval abused, so there should be no need for an extra hop as long as we include _all_ relevant swap and deposit data in the permit2 signed witness data.

Should we consider only using the extra hop for approvals, permit, and transferWithAuthorization?

Comment (nicholaspai) [reply contracts/SpokePoolV3Periphery.sol:171]: Bennett and I discussed this as well. We discussed moving swapAndBridgeWithPermit2 to the Periphery in a separate PR but since this PR only adds Permit2 functionality we should be good to put all that permit2 logic in the Periphery rather than the Proxy. @bmzig thoughts?

Comment (bmzig) [reply contracts/SpokePoolV3Periphery.sol:171]: Yeah I think this works. I agree that we will need to refactor this in general once the correct functionality to the non-approval based flows is finished.

Comment (nicholaspai) [reply contracts/SpokePoolV3Periphery.sol:171]: I've moved the permit and auth functions back to the periphery, which reduced the code diff a lot 

https://github.com/across-protocol/contracts/pull/777/files

File: contracts/SpokePoolV3Periphery.sol
Comment (nicholaspai) [inline contracts/SpokePoolV3Periphery.sol:424]: Can you not infer the transferDetails from the depositData or swapData?

File: contracts/SpokePoolV3Periphery.sol
Comment (nicholaspai) [reply contracts/SpokePoolV3Periphery.sol]: i'm not sure I understand this comment, shouldn't you include the types here if we need to get this to get permit2 to work?

Comment (bmzig) [reply contracts/SpokePoolV3Periphery.sol]: You're right. This will be added here. I don't think this would revert per se, but right now it is vulnerable to griefing, since the swapData is not part of the permit witness.

File: contracts/SpokePoolV3Periphery.sol
Comment (bmzig) [reply contracts/SpokePoolV3Periphery.sol]: Yes they are. Good call. I added it here https://github.com/across-protocol/contracts/pull/782/commits/e5473c57e07ab0a517ab11e93b819265a4e7ec5e.

The commit here was pretty large partially due to adding the hashing functions needed for EIP712.

File: contracts/SpokePoolV3Periphery.sol
Comment (bmzig) [reply contracts/SpokePoolV3Periphery.sol]: Added here. I needed to refactor the structs since there are some overlaps between the output tokens in swaps and the input tokens in deposits. Also, since we need to include all the swap and deposit data in the permit2 witness, I had to move some arguments which were formerly not in the `SwapData/DepositData` struct into those structs so that they would be included in the witness (things like `inputToken/inputAmount`). This was done here https://github.com/across-protocol/contracts/pull/782/commits/e5473c57e07ab0a517ab11e93b819265a4e7ec5e

File: contracts/SpokePoolV3Periphery.sol
Comment (bmzig) [inline contracts/SpokePoolV3Periphery.sol:769]: Thoughts on moving these hashing functions and `DepositData/BaseDepositData/SwapAndDepositData` and all EIP712 stuff to a library? Something like PeripherySigningLib?

Comment (nicholaspai) [reply contracts/SpokePoolV3Periphery.sol:769]: yes please!

Comment (nicholaspai) [reply contracts/SpokePoolV3Periphery.sol:769]: This contract is getting too big

Comment (bmzig) [reply contracts/SpokePoolV3Periphery.sol]: Library added here: https://github.com/across-protocol/contracts/pull/782/commits/e1162402065398e41221f13ffb2a25a39f2d03ef

File: contracts/SpokePoolV3Periphery.sol
Comment (bmzig) [reply contracts/SpokePoolV3Periphery.sol]: Good call. I checked with forge and it seems like we also save ~800 gas if we do this. https://github.com/across-protocol/contracts/pull/782/commits/e1162402065398e41221f13ffb2a25a39f2d03ef

File: contracts/SpokePoolV3Periphery.sol
Comment (nicholaspai) [reply contracts/SpokePoolV3Periphery.sol]: I wouldn't have moved the Swap/Deposit data into the "Signing" lib, i would have kept those in the Periphery where they are used. Just a nit

Comment (bmzig) [reply contracts/SpokePoolV3Periphery.sol]: I'll change it.

File: contracts/SpokePoolV3Periphery.sol
Comment (nicholaspai) [inline contracts/SpokePoolV3Periphery.sol:586]: Can you add unit tests for these three `TransferType` flows? I also think we should move `TransferType` into this contract as its not really a "signing" related struct

**Review by nicholaspai — APPROVED** [review]
let's just get unit tests in especially for any behavior in _swapAndBridge. I'm more ok with not having unit tests yet for the external permit/auth/permit2 logic yet but I want to make sure the swapAndBridge() and deposit() functions work well

File: contracts/SpokePoolV3Periphery.sol
Comment (nicholaspai) [inline contracts/SpokePoolV3Periphery.sol:590]: nice that the test caught this

File: contracts/SpokePoolV3Periphery.sol
Comment (nicholaspai) [inline contracts/SpokePoolV3Periphery.sol:591]: And this, unit tests doing their job!

File: contracts/SpokePoolV3Periphery.sol
Comment (nicholaspai) [inline contracts/SpokePoolV3Periphery.sol:590]: This third TransferType shouldn't be called `EIP1271Signature` i don't think, I think its more appropriately named something to do with Permit2. Technically, we could use another method that requires checking the Periphery's EIP1271Signature that doesn't involve Permit2, right?

Comment (bmzig) [reply contracts/SpokePoolV3Periphery.sol:590]: Yeah that's correct. We could use it for other things. I renamed it to `Permit2Approval` instead here https://github.com/across-protocol/contracts/pull/782/commits/f77331676ed7058639c70429ba370b7db9359112

**Pull request merged**
